static/js/favicon.js

```diff
@@ -52,14 +59,49 @@ export function update_favicon(new_message_count, pm_count) {
                 ? `${Math.floor(new_message_count / 1e3 ** pow)}${suffix}`
                 : "∞";
         const count_long = count.length > 2;
-        const rendered_favicon = render_favicon_svg({
-            count,
-            count_long,
-            have_pm: pm_count !== 0,
-            favicon_font_url,
-        });
 
-        load_and_set_favicon(rendered_favicon);
+        // we need to generate a data uri, we cannot just pass a href into
+        // the template because we generate an immutable blob from the
+        // template which would not include the rendered image
+        if (user_settings.realm_icon_as_favicon && page_params.realm_icon_url) {
+            if (!image_data_URI || should_update_image_data_URI) {
+                const reader = new FileReader();
+                // readAsDataURL converts the blob to base64 and calls onload
+                fetch(page_params.realm_icon_url)
+                    .then((response) => response.blob())
+                    .then((blob) => reader.readAsDataURL(blob));
+                reader.addEventListener("load", () => {
+                    // store image_data_URI
+                    image_data_URI = reader.result;
```

Comment:
Always prefer `async`/`await` over `.then`.

```js
const response = await fetch(page_params.realm_icon_url);
reader.readAsDataURL(await response.blob());
await new Promise((resolve) => {
    reader.addEventListener("load", () => {
        resolve();
    });
});
icon_data_url = reader.result;
```

This will also let you fix the duplicated code below (since you can conditionally `await`).